clients/gtest/rocblas_gtest_main.cpp

```diff
@@ -159,46 +159,70 @@ static void rocblas_set_listener()
 /*********************************************
  * Signal-handling for detecting test faults *
  *********************************************/
-// Id of the thread which is catching signals
-static volatile pthread_t rocblas_test_sighandler_tid;
 
-// sigjmp_buf describing stack frame to go back to
-static sigjmp_buf rocblas_sigjmp_buf;
+// State of the signal handler
+static struct
+{
+    // Whether rocblas_sigjmp_buf is set and catching signals is enabled
+    volatile sig_atomic_t rocblas_sighandler_enabled;
+
+    // Id of the thread which is catching signals
+    volatile pthread_t rocblas_test_sighandler_tid;
+
+    // sigjmp_buf describing stack frame to go back to
+    sigjmp_buf rocblas_sigjmp_buf;
+} state;
 
-// Whether rocblas_sigjmp_buf is set and catching signals is enabled
-static volatile sig_atomic_t rocblas_sighandler_enabled = false;
+// The signal which was received
+static volatile int signal_received;
 
-// Signal handler
+// Number of seconds each test is allowed to take before all testing is killed.
+// If this time limit is exceeded for any particular test, it is assumed that
+// there is a permanent deadlock, and testing is terminated.
+constexpr int TEST_TIMEOUT = 600;
+
+// Signal handler (must have external "C" linkage)
 extern "C" void rocblas_test_signal_handler(int sig)
 {
-    int e = errno; // Save errno
+    int saved_errno = errno; // Save errno
+
+    // If this is an alarm timeout, we abort
+    if(sig == SIGALRM)
+    {
+        static constexpr char msg[]
+            = "\nAborting tests due to an alarm timeout.\n\n"
+              "This could be due to a deadlock caused by mutexes being left locked\n"
+              "after a previous test's signal was caught and partially recovered from.\n";
+        write(STDERR_FILENO, msg, sizeof(msg) - 1);
+        rocblas_abort();
+    }
```

Comment:
When more than 10 minutes have elapsed, all tests are aborted, and this message is displayed.

Hangs may occur if a signal or exception occurs during a test and a mutex is left locked. This alarm catches that and prevents hanging tests.